decidim-core/app/services/decidim/truncation.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+module Decidim
+  class Truncation
+    include ActionView::Context
+    include ActionView::Helpers::TagHelper
+
+    # Truncates text and html content
+    # text - Content to be truncated
+    # options - Hash with the options
+    #         max_length: An Integer maximum number of characters
+    #         tail: A string suffix to be added after truncation
+    #         count_tags: A boolean if html is calculate to max length, otherwise just content
+    #         count_tail: A boolean value that determines whether max_length contains the tail
+    #         tail_before_final_tag: A boolean, show tail inside of tag where text is cutted or before final closing tag.
+    def initialize(text, options = {})
+      @options = {
+        max_length: options[:max_length] || 30,
+        tail: options[:tail] || "...",
+        count_tags: options[:count_tags] || false,
+        count_tail: options[:count_tail] || false,
+        tail_before_final_tag: options[:tail_before_final_tag] || false
+      }
+      @document = Nokogiri::HTML::DocumentFragment.parse(text)
+      @tail_added = false
+      @remaining = initial_remaining
+    end
+
+    # Truncate text or html content added in constructor
+    # Returns truncated html
+    def truncate
+      content_array = []
+
+      document.children.each do |node|
+        if node_length(node) > @remaining
+          content_array << truncate_last_node(node)
+          break
+        end
+
+        change_quotes(node)
+        content_array << node.to_html
+        @remaining -= node_length(node)
+      end
+
+      # Nokogiri's to_html escapes &quot; to &amp;quot; and we do not want extra &amp so we have to unescape.
+      CGI.unescape_html content_array.join.html_safe
+    end
+
+    private
+
+    attr_accessor :tail_added, :remaining
+    attr_reader :document, :options
+
+    def change_quotes(node)
```

Comment:
Why is this necessary?

Reply:
It made it work more like the old solution. For example here: https://github.com/decidim/decidim/blob/f3118f22338cb77480e2f46e4f08bd2dab47df9f/decidim-proposals/spec/helpers/map_helper_spec.rb#L63 . But I removed it for now, because not sure if it's needed.

Reply:
Apparently we have to also escape HTML from the content, so replaces this with ```escape_html_from_content(node)```.